### PR TITLE
Lazy Mamba extra-buffer allocation for radix cache

### PR DESCRIPTION
**Pull request by YAMY1234**
## Summary

Co-authored-by: @hanming-lu

This PR implements lazy Mamba extra-buffer allocation for radix-cache tracking. It keeps the existing request-owned ping-pong buffer representation for backward compatibility, while reducing fixed extra-buffer allocation by materializing the second track slot only when a tracking boundary may need it.

The optimization is motivated by the memory-overhead reduction explored in #22542. This PR preserves that memory-saving goal, but keeps the existing ping-pong lifecycle instead of replacing it with a new pending-slot ownership model. This keeps the change local, additive, and closer to the current request-owned buffer semantics.

Concretely:

- Lazy mode allocates one Mamba ping-pong track slot at request allocation time and leaves the other entry as `-1`.
- At a Mamba track boundary, the missing target slot is allocated on demand.
- If the target slot already exists, it is reused. This matters for speculative decoding where the same boundary may be prepared before the final accept length is known.
- If target-slot allocation fails under pressure, this PR does not evict radix cache just to create a transient boundary slot. It falls back to a detached keep-slot track and skips unsafe finished-request Mamba insertion when needed.
- The non-lazy/full ping-pong path remains unchanged.

## Design Rationale

The design keeps the existing request-owned ping-pong buffer lifecycle for compatibility. That allows the memory optimization to be implemented as a local lazy-allocation change instead of a broader ownership rewrite.

The main semantic cleanup is that `mamba_next_track_idx` consistently means the next write target. In lazy mode, the request usually owns one valid keep slot and one missing target slot (`-1`); the missing target is materialized only near a track boundary.

## Validation

Unless otherwise noted:

- Model: `Qwen/Qwen3-Next-80B-A3B-Instruct`
- `N = max_running_requests = 32`
- `mamba_track_interval = 128`
- Radix cache enabled
- `5N` means `max_mamba_cache_size = 160`
- KL is computed only on rows that hit decode cache, which checks that a cache hit uses a valid Mamba radix state.

### Accuracy / KL

| Case | Job | Decode-cache hit | Cached tokens | Decode-cache KL | Result |
|---|---:|---:|---|---:|---|
| no-MTP, `N=32`, `5N` | `1961806` | `32/32` | all `896` | `0.0015160608971137046` | pass `< 0.008` |
| no-MTP stress, `64 req = 2N`, `5N` | `1961808` | `31/64` | `0` or `896` | `0.001059574473602977` | pass `< 0.008` on hit rows |
| MTP, `N=32`, `5N` | `1961809` | `32/32` | `896` or `1024` | `0.0008251577837979901` | pass `< 0.008` |
| MTP debug rerun, same config | `1961859` | `32/32` | `896` or `1024` | `0.0016668516675800381` | pass `< 0.008` |

### Cache-Hit Comparison

| Workload | Baseline old ping-pong | Pending-slot PR #22542 | This PR |
|---|---:|---:|---:|
| no-MTP, `N=32`, `5N` | `31/32` | `32/32` | `32/32` |
| MTP, `N=32`, `5N` | `31/32` | `32/32` | `32/32` |
| no-MTP stress, `64 req = 2N`, `5N` | `0/64` | `31/64` | `31/64` |

This PR recovers the cache opportunity of the pending-slot implementation while keeping the ping-pong buffer layout.

### Boundary / Pool-Pressure Debug Counters

Counters below are from runs with gated debug logging enabled.

| Event | no-MTP N32 `1961806` | no-MTP stress `1961808` | MTP debug `1961859` |
|---|---:|---:|---:|
| `target_alloc_failed` | `464` | `1504` | `0` |
| `fallback_detach_keep` | `464` | `1504` | `0` |
| `fallback_no_keep` | `0` | `0` | `0` |
| `target_alloc_success` | `560` | `544` | `1024` |
| `target_exists` | `0` | `0` | `756` |
| `cache_finished_skip_mamba` | `0` | `128` | `0` |
| `evict_begin` | `0` | `136` | `0` |
| `evict_node` | `0` | `428` | `0` |
| `cache_finished_insert` | `396` | `520` | `396` |
| `release_detached_track` | `0` | `128` | `0` |
| `restore_detached_track` | `464` | `1376` | `0` |

Key readout:

- In the no-MTP `N=32` run, boundary target allocation failed 464 times, but this PR used detached keep-slot fallback without any schedule-side radix eviction, and still achieved `32/32` hits.
- In the stress run, evictions still happen later due to radix insert pressure, but they are no longer caused by retrying target-slot allocation at the tracking boundary.
- In the MTP run, all target slots were either allocated or already present, and cache-hit/KL behavior was normal.

## Local Checks

- `python3 -m compileall -q` on changed Python files
- `pre-commit run --files` on changed Python files

---
### CI States

Latest PR Test (Base): :x: **Missing `run-ci` label** -- add it to run CI tests.
Latest PR Test (Extra): :x: **Blocked** -- `run-ci` is required first.

### CODE REVIEW

File: python/sglang/srt/managers/scheduler_components/batch_result_processor.py [hunk 1]
```diff
@@ -850,33 +856,116 @@ def _mamba_prefix_cache_update(
...
+            batch.mamba_track_mask is None
+            or batch.mamba_track_indices is None
+            or not bool(batch.mamba_track_mask[i].item())
+            or req.mamba_ping_pong_track_buffer is None
+        ):
+            return False
+        track_slot = batch.mamba_track_indices[i].item()
+        if track_slot == -1:
+            return False
+        return not bool((req.mamba_ping_pong_track_buffer == track_slot).any().item())
```
[the remark this answers is not in the document]
Comment (YAMY1234) [reply python/sglang/srt/managers/scheduler_components/batch_result_processor.py]: Addressed in 90d30ac5e. Instead of adding a req-level CPU mirror for the ping-pong buffer, the patch keeps the bookkeeping local to the lazy batch plan: `mamba_lazy_track_mask_cpu` and `mamba_lazy_detached_track_cpu`. The result processor now uses these CPU-side lazy-plan fields for prepared/detached decisions and no longer performs per-request GPU scalar reads for those checks.

File: python/sglang/srt/managers/schedule_batch.py [hunk 2]
```diff
@@ -1463,6 +1463,14 @@ def set_mamba_track_indices_from_reqs(batch):
     )
 
 
+def set_mamba_track_indices_from_slots(batch, track_slots: List[int]):
+    batch.mamba_track_indices = (
+        torch.tensor(track_slots, dtype=torch.int64)
+        .pin_memory()
+        .to(device=batch.device, non_blocking=True)
+    )
```
[the remark this answers is not in the document]
Comment (YAMY1234) [reply python/sglang/srt/managers/schedule_batch.py]: Addressed in 90d30ac5e. The result processor no longer reads `batch.mamba_track_indices[i].item()` for lazy detached restore; it passes the tensor element directly when writing the slot back. This avoids adding a broader `mamba_track_indices_cpu` mirror for the non-lazy path.

File: python/sglang/srt/managers/schedule_batch.py [hunk 3]
```diff
@@ -2494,19 +2563,33 @@ def prepare_for_decode(self):
...
+                    )
+                    set_mamba_track_indices_from_slots(self, track_slots)
+                else:
+                    track_mask_cpu = interval_mask_cpu
+                    set_mamba_track_indices_from_reqs(self)
+                self.mamba_track_mask = (
+                    torch.tensor(track_mask_cpu, dtype=torch.bool)
+                    .pin_memory()
+                    .to(device=self.device, non_blocking=True)
+                )
```
[the remark this answers is not in the document]
Comment (YAMY1234) [reply python/sglang/srt/managers/schedule_batch.py]: Addressed in 90d30ac5e with a narrower lazy-only field: `mamba_lazy_track_mask_cpu`. This keeps the CPU-side plan only for the lazy extra-buffer path that needs result-side ownership bookkeeping.

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 4]
```diff
@@ -648,6 +646,72 @@ def get_mamba_ping_pong_other_idx(self, mamba_next_track_idx: int) -> int:
...
+            "try to increase --mamba-full-memory-ratio."
+        )
+        buf = torch.full(
+            (self.mamba_ping_pong_track_buffer_size,),
+            -1,
+            dtype=slots.dtype,
+            device=slots.device,
+        )
+        buf[:n] = slots
+        req.mamba_ping_pong_track_buffer = buf
```
[the remark this answers is not in the document]
Comment (YAMY1234) [reply python/sglang/srt/mem_cache/memory_pool.py:680]: Addressed in 90d30ac5e without adding `req.mamba_ping_pong_track_buffer_cpu`. The req-owned device buffer remains the authoritative state; detached ownership is tracked by the batch-local lazy plan instead, avoiding another req-level mirror to keep in sync.

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 5]
```diff
@@ -648,6 +646,72 @@ def get_mamba_ping_pong_other_idx(self, mamba_next_track_idx: int) -> int:
...
+        """Update a ping-pong slot value and sync the device-side mapping.
+
+        The req holds the authoritative buffer; this keeps the
+        req_index_to_mamba_ping_pong_track_buffer_mapping in sync so that
+        set_mamba_track_indices_from_reqs reads correct slot indices.
+        """
+        req.mamba_ping_pong_track_buffer[idx] = value
+        self.req_index_to_mamba_ping_pong_track_buffer_mapping[req.req_pool_idx] = (
+            req.mamba_ping_pong_track_buffer
+        )
```
[the remark this answers is not in the document]
Comment (YAMY1234) [reply python/sglang/srt/mem_cache/memory_pool.py]: Addressed in 90d30ac5e by avoiding the req-level CPU mirror entirely. Since `req.mamba_ping_pong_track_buffer_cpu` is not introduced, `set_mamba_ping_pong_slot` does not need extra mirror synchronization.
